backend/hitas/oracle_migration/runner.py

```diff
@@ -847,3 +876,218 @@ def create_codes(
+                    bulk_owners.append(new_owner)
+                    buyers.append(new_owner)
+        return buyers
+
+    def create_ownerships(buyers: list[Owner], is_latest_sale) -> None:
+        new_ownerships = [
+            Ownership(
+                apartment=v,
+                owner=buyer,
+                sale=new,
+                percentage=100 / len(buyers),  # Assume equal split, as there is no way to know the real percentages
+                deleted=is_latest_sale and None or timezone.now(),  # We only want the latest sale to be non-deleted
+            )
+            for buyer in buyers
+        ]
+        bulk_ownerships.extend(new_ownerships)
+
+    count = 0
+    bulk_apartment_sales = []
+    bulk_owners = []
+    bulk_ownerships = []
+
+    for apartment_oracle_id, v in converted_data.apartments_by_oracle_id.items():
+        # Edge case, skip importing sales.
+        # These are apartments with missing data, incomplete apartments or rented apartments.
+        if not v.first_purchase_date:
+            continue
+
+        # Apartment has been bought, but never resold.
+        # All necessary data to create a sale is in apartment fields (and no sales records should even exist).
+        if v.first_purchase_date and not v.latest_purchase_date:
+            create_sale_for_original_ownership()
+            count += 1
+            continue
+
+        created_apartment_sales: dict[str, set[OwnerKey]] = {}  # Keep track of owners in this apartments sales
+        current_owner_keys = {OwnerKey(o.owner.identifier, o.owner.name) for o in v.ownerships.all()}
+        is_latest_sale_created = False  # Used to determine if creating a sale based only on apartment data is needed.
+
+        # Apartment has been resold, so fetch all sales (=confirmed max price calculations).
+        apartment_sales = get_all_apartment_sales(apartment_oracle_id)
+        for sale in apartment_sales:
+            # Skip any calculations with missing buyer or purchase date, as it is most likely
+            # only a calculation, and couldn't be imported anyway due to missing data.
+            # The only exception to this rule is the latest calculation (the active one).
+            if sale["monitoring_state"] != ApartmentSaleMonitoringState.ACTIVE.value and (
+                not sale["purchase_date"] or not sale["buyer_name_1"] or sale["buyer_name_1"] == "Ei tiedossa"
+            ):
+                continue
+
+            # Special logic for the last active calculation if its missing data.
+            if sale["monitoring_state"] == ApartmentSaleMonitoringState.ACTIVE.value and not sale["purchase_date"]:
+                # Check if the calculation was at valid the last time apartment was sold.
+                if not sale["purchase_date"]:
+                    alt_date = sale["calculation_date"] or sale["notification_date"]
+                    if not (alt_date <= v.latest_purchase_date <= (alt_date + relativedelta(months=4))):
```

Comment:
Where do you get the value +4 months?

Reply:
4 Months is the longest a calculation can be active (when Surface Area Price Ceiling, otherwise it is 3 months)